Add isort.force-sort-within-sections setting

mattoberle opened the pull request:
This commit is a first attempt at addressing issue #1003.

The default `isort` behavior is `force-sort-within-sections = false`,
which places `from X import Y` statements after `import X` statements.

When `force-sort-within-sections = true` all imports are sorted by module name.
When module names are equivalent, the `import` statement comes before the `from` statement.
- https://pycqa.github.io/isort/docs/configuration/options.html#force-sort-within-sections

~Initially I tried out an `enum` + `match` approach but I was struggling to come up with anything that didn't require `Copy` support.~

---

~The approach in this PR _works_ from what I've been able to test...
But it's lacking in clarity.
The `merge_imports` function returns a vector of tuples that represent the final import order:~

```
[(0 or 1, idx)]
```

~Where...~
- ~`0` means use the `import_block.import` vector with index `idx`~
- ~`1` means use the `import_block.import_from` vector with index `idx`~

~Those integers are essentially a stand-in for a type or enum, I'm just not sure what the idiomatic thing to return would be.~

- ~An new enum like `ImportType::ImportFrom / ImportType::Import` defined in `types.rs`?~
- ~Is there a better way to do this lazily (eg. return an iterator) rather than eagerly construct the vector in `merge_imports`?~

**Edit:** Swapped out the previous approach with an iterator and enum match.

mattoberle commented on src/isort/sorting.rs:
```diff
@@ -69,3 +69,17 @@ pub fn cmp_import_from(import_from1: &ImportFromData, import_from2: &ImportFromD
     })
 }
+
+/// Compare two `AnyImport` enums which may be `Import` or `ImportFrom` structs.
+pub fn cmp_any_import(a: &AnyImport, b: &AnyImport) -> Ordering {
+    match (&a, &b) {
+        (AnyImport::Import(x), AnyImport::Import(y)) => cmp_modules(&x.0, &y.0),
+        (AnyImport::ImportFrom(x), AnyImport::Import(y)) => {
+            natord::compare_ignore_case(x.0.module.unwrap(), y.0.name)
+        }
+        (AnyImport::Import(x), AnyImport::ImportFrom(y)) => {
+            natord::compare_ignore_case(x.0.name, y.0.module.unwrap())
+        }
+        (AnyImport::ImportFrom(x), AnyImport::ImportFrom(y)) => cmp_import_from(&x.0, &y.0),
+    }
+}
```
Guessing that `.unwrap()` could be unsafe here, although I don't understand the case that results in `module` being `None`.

andersk replied:
That happens for `from . import foo` (or `..`, `...`, etc.).

mattoberle replied:
Makes sense.
I threw one of those into my test cases but I suppose that since the block in the case is only homogeneous `import from` statements we never hit these middle conditions.

mattoberle replied:
I dug into this a bit.

Given the current callers, `.unwrap()` will never return `None` because relative imports _always_ use the `from X import Y` syntax. An import like `import .my_module` is invalid.
Relative imports are sorted in their own section and aren't compared to other first-party imports.

That doesn't necessarily guarantee this code path won't be hit with a missing module given a different set of settings down the road.

What's the best way to handle this, something like?
```rs
fn cmp_option_str(a: Option<&String>, b: Option<&String>) -> Ordering {
    match (a, b) {
        (None, None) => Ordering::Equal,
        (None, Some(_)) => Ordering::Less,
        (Some(_), None) => Ordering::Greater,
        (Some(x), Some(y)) => natord::compare_ignore_case(x, y),
    }
}
```

charliermarsh replied:
Yeah we should do something like what you have there, just to be safe. You can see some examples of that `None` handling in functions like `cmp_level`.

charliermarsh commented:
Will review this today, thank you!

charliermarsh commented on src/isort/mod.rs:
```diff
@@ -585,29 +586,49 @@ pub fn format_imports(
+        for import in it {
+            match import {
+                AnyImport::Import(x) => {
+                    let (alias, comments) = &x;
```
Nit: can you destructure directly? (Like: `AnyImport::Import((alias, comments))`?)

charliermarsh commented on src/isort/mod.rs:
```diff
@@ -585,29 +586,49 @@ pub fn format_imports(
+            )
+            .sorted_by(|a, b| {
+                if force_sort_within_sections {
+                    cmp_any_import(a, b)
```
Maybe instead, `let mut it`, and then call `sorted_by` on it if `force_sort_within_sections`? That way, we avoid the oddity of `else { Ordering::Greater }`.

mattoberle replied:
I was sort of able to do this, although I had to `.collect` before (and within) the `if` block to ensure `it ` was the same type.

charliermarsh replied:
If we use `sort_by` instead of `sorted_by`, we can avoid that allocation.

mattoberle commented:
Okay, I think I have this (nearly) passing (maybe a few tasks to get it up-to-date with upstream).

- Modified some match blocks to deref in the patterns.
- Put the additional sort into an `if` block.
- Replaced the bare `.unwrap` with something safer.

I was struggling to put together a function that would take `Option<&str>` arguments for a comparison.
`alias.name` is always the same type, but `import_from.module` seems to be `Option<String>` or `Option<str>` depending on whether I was running in GHA or locally. (Maybe something to do with the `wasm` target)?

I ultimately just unwrapped to an empty str, which I think results in the same set or sort rules getting applied (`""` is before any non-empty string).

---

In the broader picture: Is it weird to take something called `OrderedImportBlock` and then re-order it?
That was definitely the easy-way, but I started another branch that refactored things a bit...

1. Collect all import data and put them into an unsorted `Vec<EitherImport>` (or unevaluated iterator).
2. Apply sort rules.
3. Iterate over the `EitherImport`s, writing to output.

In that pattern there is no  concept of `OrderedImportBlock` with different `import` and `import_from` vectors.
Sorting the statements is contained within the `cmp_*` function and match blocks can consider the various settings.

charliermarsh commented:
I'd definitely be interested to see your other branch!

charliermarsh commented:
(I can help with the Option<String> conversions, we changed those recently.)

mattoberle commented:
> Merging this for now! I see what you're saying with the refactor. I may knock it out, you're also welcome to PR it. Thank you for the contribution :)

Thanks for your help with this!
My existing branch on the refactor is incomplete at this point (and out-of-date WRT feedback here).
If I find some time this week I'll work on getting it PR-ready, but if you have a vision don't let that stop you-- I think the exercise will still be worth it on my end.
